## 38. 프로젝트 루트 확인

In [343]:
from pathlib import Path

project_root = Path.cwd()

if project_root.name == "notebooks":
    project_root = project_root.parent

data_dir = project_root / "data" / "raw"

print("프로젝트 루트:", project_root)
print("데이터 폴더:", data_dir)
print("데이터 폴더 존재:", data_dir.exists())

프로젝트 루트: c:\dev\ai-data-analysis
데이터 폴더: c:\dev\ai-data-analysis\data\raw
데이터 폴더 존재: True


## 39. 필수 파일 확인

In [344]:
required_files = [
    "customers.csv",
    "products.csv",
    "orders.csv",
    "order_items.csv",

]
for file_name in required_files:
    file_path = data_dir / file_name
    print(file_name, file_path.exists(), file_path.stat().st_size if file_path.exists() else 0)

customers.csv True 5641
products.csv True 4103
orders.csv True 11655
order_items.csv True 15356


In [345]:
import pandas as pd

customers = pd.read_csv(data_dir / "customers.csv")
products = pd.read_csv(data_dir / "products.csv")
orders = pd.read_csv(data_dir / "orders.csv")
order_items = pd.read_csv(data_dir / "order_items.csv")

## 41. DataFrame 사전 만들기

In [346]:
datasets = {

    "customers": customers,
    "products": products,
    "orders": orders,
    "order_items": order_items,
}


## 42. 크기 요약

In [347]:
shape_summary = pd.DataFrame(
    [
        {
            "dataset": name,
            "rows": df.shape[0],
            "columns": df.shape[1],
        }
        for name, df in datasets.items()
    ]
)

print(shape_summary)


       dataset  rows  columns
0    customers   150        6
1     products   100        4
2       orders   300        5
3  order_items   765        5


## 데이터셋별 한 행의 의미

- customers: 고객 한 명
- products: 상품 한 개
- orders: 주문 한 건
- order_items: 주문에 포함된 상품 한 항목

## 44. 처음과 마지막 행 확인


In [348]:
for name, df in datasets.items():
    print(f"\n[{name}] 앞부분")
    display(df.head(10))
    print(f"[{name}] 뒷부분")
    display(df.tail(10))



[customers] 앞부분


,customer_id,name,gender,age,city,signup_date
0,1,김수민,F,19,광주,2024-07-15
1,2,김정호,F,32,대구,2025-11-28
2,3,이경수,F,61,성남,2024-07-08
3,4,조영호,F,55,울산,2026-05-09
4,5,이예원,F,19,부산,2024-10-09
5,6,김지원,F,32,성남,2026-07-23
6,7,이상현,F,53,인천,2025-01-07
7,8,이성호,M,32,수원,2025-01-19
8,9,송지민,M,69,서울,2025-11-14
9,10,유도현,F,62,울산,2024-11-01


[customers] 뒷부분


,customer_id,name,gender,age,city,signup_date
140,141,김진호,M,40,광주,2023-10-05
141,142,하서연,F,32,서울,2023-09-04
142,143,김예은,F,43,대전,2024-01-28
143,144,김정자,M,22,광주,2025-02-27
144,145,조정호,M,59,성남,2024-01-28
145,146,김숙자,M,61,성남,2026-01-18
146,147,이정남,M,19,부산,2025-03-09
147,148,오도현,M,29,고양,2026-07-11
148,149,김정자,M,20,부산,2024-11-14
149,150,조미영,M,40,대전,2025-12-30



[products] 앞부분


,product_id,product_name,category,price
0,1,전자기기 상품 001,전자기기,160000
1,2,도서 상품 002,도서,34000
2,3,전자기기 상품 003,전자기기,152000
3,4,생활용품 상품 004,생활용품,70000
4,5,식품 상품 005,식품,186000
5,6,전자기기 상품 006,전자기기,5000
6,7,도서 상품 007,도서,142000
7,8,스포츠 상품 008,스포츠,189000
8,9,스포츠 상품 009,스포츠,193000
9,10,스포츠 상품 010,스포츠,55000


[products] 뒷부분


,product_id,product_name,category,price
90,91,스포츠 상품 091,스포츠,59000
91,92,식품 상품 092,식품,111000
92,93,전자기기 상품 093,전자기기,89000
93,94,도서 상품 094,도서,124000
94,95,전자기기 상품 095,전자기기,20000
95,96,생활용품 상품 096,생활용품,112000
96,97,전자기기 상품 097,전자기기,154000
97,98,스포츠 상품 098,스포츠,10000
98,99,뷰티 상품 099,뷰티,200000
99,100,도서 상품 100,도서,102000



[orders] 앞부분


,order_id,customer_id,order_date,payment_method,order_status
0,1,123,2026-06-02,card,completed
1,2,77,2025-08-18,naver_pay,cancelled
2,3,138,2025-12-15,bank_transfer,cancelled
3,4,57,2026-02-25,kakao_pay,cancelled
4,5,125,2026-01-16,card,cancelled
5,6,87,2026-04-16,naver_pay,completed
6,7,120,2025-12-24,bank_transfer,refunded
7,8,137,2025-08-29,card,cancelled
8,9,145,2026-02-15,card,completed
9,10,110,2026-04-11,bank_transfer,cancelled


[orders] 뒷부분


,order_id,customer_id,order_date,payment_method,order_status
290,291,95,2025-11-29,kakao_pay,completed
291,292,24,2026-05-01,naver_pay,cancelled
292,293,149,2026-01-06,card,cancelled
293,294,147,2025-11-16,bank_transfer,completed
294,295,122,2026-02-13,bank_transfer,completed
295,296,116,2026-02-28,card,completed
296,297,22,2026-02-10,bank_transfer,completed
297,298,64,2026-01-20,naver_pay,cancelled
298,299,135,2025-09-03,bank_transfer,completed
299,300,96,2026-03-23,kakao_pay,cancelled



[order_items] 앞부분


,order_item_id,order_id,product_id,quantity,unit_price
0,1,1,100,3,102000
1,2,1,87,5,25000
2,3,1,7,3,142000
3,4,1,9,3,193000
4,5,2,72,4,189000
5,6,3,33,5,124000
6,7,3,20,3,80000
7,8,3,11,5,115000
8,9,4,45,3,62000
9,10,4,84,4,78000


[order_items] 뒷부분


,order_item_id,order_id,product_id,quantity,unit_price
755,756,295,82,5,75000
756,757,296,36,1,126000
757,758,296,40,1,174000
758,759,297,20,5,80000
759,760,297,42,4,28000
760,761,298,40,4,174000
761,762,299,8,2,189000
762,763,299,12,4,175000
763,764,300,59,5,32000
764,765,999,300,15,155000


## 45. 컬럼명 확인

In [349]:

for name, df in datasets.items():
    print(name, df.columns.tolist())


customers ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date']
products ['product_id', 'product_name', 'category', 'price']
orders ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status']
order_items ['order_item_id', 'order_id', 'product_id', 'quantity', 'unit_price']


## 46. 컬럼명 공백 점검

In [350]:
for name, df in datasets.items():
    
    suspicious = [
        col for col in df.columns
        if col != col.strip()
    ]
    print(name, suspicious)

customers []
products []
orders []
order_items []


## 47. info() 실행

In [351]:
for name, df in datasets.items():

    print(f"\n===== {name} =====")
    df.info()


===== customers =====
<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   customer_id  150 non-null    int64
 1   name         150 non-null    str  
 2   gender       150 non-null    str  
 3   age          150 non-null    int64
 4   city         150 non-null    str  
 5   signup_date  150 non-null    str  
dtypes: int64(2), str(4)
memory usage: 7.2 KB

===== products =====
<class 'pandas.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   product_id    100 non-null    int64
 1   product_name  100 non-null    str  
 2   category      100 non-null    str  
 3   price         100 non-null    int64
dtypes: int64(2), str(2)
memory usage: 3.3 KB

===== orders =====
<class 'pandas.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 5 columns):

## 48. dtype 요약표

In [352]:
dtype_rows = []

for name, df in datasets.items():
    for column, dtype in df.dtypes.items():
        dtype_rows.append(
            {
                "dataset": name,
                "column": column,
                "dtype": str(dtype),
            }

        )
dtype_summary = pd.DataFrame(dtype_rows)
dtype_summary


,dataset,column,dtype
0,customers,customer_id,int64
1,customers,name,str
2,customers,gender,str
3,customers,age,int64
4,customers,city,str
5,customers,signup_date,str
6,products,product_id,int64
7,products,product_name,str
8,products,category,str
9,products,price,int64


## 49. 날짜형 후보 확인

In [353]:
date_columns = {
    "customers": ["signup_date"],
    "orders": ["order_date"],
}

for dataset_name, columns in date_columns.items():
    df = datasets[dataset_name]
    for column in columns:
        converted = pd.to_datetime(
            df[column],
            errors="coerce"
        )

        print(
            dataset_name,
            column,
            "원본 dtype:", df[column].dtype,
            "변환 실패:", converted.isna().sum()
        )

customers signup_date 원본 dtype: str 변환 실패: 0
orders order_date 원본 dtype: str 변환 실패: 0


## 50. 결측치 개수

In [354]:
for name, df in datasets.items():
    print(f"\n[{name}]")
    display(df.isna().sum().to_frame("missing_count"))


[customers]


,missing_count
customer_id,0
name,0
gender,0
age,0
city,0
signup_date,0



[products]


,missing_count
product_id,0
product_name,0
category,0
price,0



[orders]


,missing_count
order_id,0
customer_id,0
order_date,0
payment_method,0
order_status,0



[order_items]


,missing_count
order_item_id,0
order_id,0
product_id,0
quantity,0
unit_price,0


## 51. 결측치 비율

In [355]:
missing_summaries = []

for name, df in datasets.items():

    summary = pd.DataFrame(
        {
            "missing_count": df.isna().sum(),
            "missing_rate_pct": (
                df.isna().mean() * 100
            ).round(2),
        }

    )

    summary.insert(0, "dataset", name)
    summary.insert(1, "column", summary.index)
    missing_summaries.append(summary.reset_index(drop=True))
missing_summary = pd.concat(
    missing_summaries,
    ignore_index=True
)
missing_summary

,dataset,column,missing_count,missing_rate_pct
0,customers,customer_id,0,0.0
1,customers,name,0,0.0
2,customers,gender,0,0.0
3,customers,age,0,0.0
4,customers,city,0,0.0
5,customers,signup_date,0,0.0
6,products,product_id,0,0.0
7,products,product_name,0,0.0
8,products,category,0,0.0
9,products,price,0,0.0


## 전체 행 중복

In [356]:
for name, df in datasets.items():
    print(name, "전체 중복:", df.duplicated().sum())

customers 전체 중복: 0
products 전체 중복: 0
orders 전체 중복: 0
order_items 전체 중복: 0


## 주요 ID 점검

In [357]:
primary_keys = {
    "customers": "customer_id",
    "products": "product_id",
    "orders": "order_id",
    "order_items": "order_item_id",
}
id_check_rows = []
for name, key in primary_keys.items():
    df = datasets[name]
    id_check_rows.append(
        {
            "dataset": name,
            "key": key,
            "missing": df[key].isna().sum(),
            "duplicates": df[key].duplicated().sum(),
            "unique": df[key].nunique(),
            "rows": len(df),
        }
    )
id_check_summary = pd.DataFrame(id_check_rows)
id_check_summary


,dataset,key,missing,duplicates,unique,rows
0,customers,customer_id,0,0,150,150
1,products,product_id,0,0,100,100
2,orders,order_id,0,0,300,300
3,order_items,order_item_id,0,0,765,765


## 54. 발견 내용을 기록합니다

## 데이터 품질 발견 기록

| 데이터셋 | 컬럼 | 문제 종류 | 건수 | 해석 | 처리 여부 |
|---|---|---|---:|---|---|
| customers | customer_id | 없음 | 0건 | 가입일자 이상없음 | N/A |
| products | product_id | 없음 | 0건 | 기본 키 무결성 정상 (결측/중복 없음) | N/A |
| orders | order_id | 없음 | 0건 | 주문일자 이상없음 | N/A |
| order_items | order_item_id | 없음 | 0건 | 기본 키 무결성 정상 (결측/중복 없음) | N/A |

In [358]:
# 55. 숫자형 기본 통계

print(customers[["age"]].describe())
print(products[["price"]].describe())
print(order_items[["quantity", "unit_price"]].describe())


              age
count  150.000000
mean    42.086667
std     15.613166
min     19.000000
25%     29.000000
50%     40.000000
75%     57.000000
max     69.000000
               price
count     100.000000
mean   110040.000000
std     56433.910574
min      5000.000000
25%     65750.000000
50%    112000.000000
75%    161000.000000
max    200000.000000
         quantity     unit_price
count  765.000000     765.000000
mean     3.069281  108622.222222
std      1.474623   56984.197109
min      1.000000    5000.000000
25%      2.000000   62000.000000
50%      3.000000  111000.000000
75%      4.000000  161000.000000
max     15.000000  200000.000000


In [359]:
 

#56. 범주형 값 확인

category_columns = {
    "customers": ["gender", "city"],
    "products": ["category"],
    "orders": ["payment_method", "order_status"],
}
for dataset_name, columns in category_columns.items():
    df = datasets[dataset_name]
    for column in columns:
        print(f"\n[{dataset_name}.{column}]")
        display(
            df[column]
            .value_counts(dropna=False)
            .to_frame("count")
        )


[customers.gender]


,count
gender,
F,84
M,66



[customers.city]


,count
city,
성남,21
광주,17
부산,16
대구,15
서울,15
울산,14
인천,14
대전,14
수원,13



[products.category]


,count
category,
스포츠,19
전자기기,17
생활용품,16
뷰티,16
도서,14
패션,11
식품,7



[orders.payment_method]


,count
payment_method,
kakao_pay,79
naver_pay,77
bank_transfer,74
card,70



[orders.order_status]


,count
order_status,
completed,184
cancelled,64
refunded,52


In [360]:
#57. 날짜 범위 확인

signup_dates = pd.to_datetime(
    customers["signup_date"],
    errors="coerce"
)
order_dates = pd.to_datetime(
    orders["order_date"],
    errors="coerce"
)

print("가입일 변환 실패:", signup_dates.isna().sum())
print("가입일 범위:", signup_dates.min(), "~", signup_dates.max())
print("주문일 변환 실패:", order_dates.isna().sum())
print("주문일 범위:", order_dates.min(), "~", order_dates.max())


가입일 변환 실패: 0
가입일 범위: 2023-08-09 00:00:00 ~ 2026-07-23 00:00:00
주문일 변환 실패: 0
주문일 범위: 2025-08-04 00:00:00 ~ 2026-08-03 00:00:00


## 58. 날짜 관계 점검 준비

주문일과 고객 가입일을 직접 비교하려면 두 데이터를 연결해야 합니다.

이번 장에서는 구조와 키가 연결 가능한지 확인하고, 실제 병합과 분석은 이후 장에서 본격적으로 진행합니다.


In [361]:
#59. 참조하지 못한 고객 ID

unknown_customer_ids = sorted(
    set(orders["customer_id"].dropna())
    - set(customers["customer_id"].dropna())
)
print("존재하지 않는 customer_id:", unknown_customer_ids)

존재하지 않는 customer_id: []


In [362]:
#60. 참조하지 못한 주문 ID

unknown_order_ids = sorted(
    set(order_items["order_id"].dropna())
    - set(orders["order_id"].dropna())
)
print("존재하지 않는 order_id:", unknown_order_ids)

존재하지 않는 order_id: [999]


In [363]:
#61. 참조하지 못한 상품 ID

unknown_product_ids = sorted(

    set(order_items["product_id"].dropna())

    - set(products["product_id"].dropna())

)
print("존재하지 않는 product_id:", unknown_product_ids)


존재하지 않는 product_id: [300]


In [364]:
#62. 반복 점검 함수

 

def inspect_dataframe(
    name: str,
    df: pd.DataFrame,
    primary_key: str | None = None,

) -> dict:
    result = {
        "dataset": name,
        "rows": df.shape[0],
        "columns": df.shape[1],
        "column_names": df.columns.tolist(),
        "dtypes": {
            column: str(dtype)
            for column, dtype in df.dtypes.items()
        },
        "missing_total": int(df.isna().sum().sum()),
        "duplicate_rows": int(df.duplicated().sum()),
    }
    if primary_key is not None:
        result["primary_key"] = primary_key
        result["key_missing"] = int(
            df[primary_key].isna().sum()
        )
        result["key_duplicates"] = int(

            df[primary_key].duplicated().sum()

        )

        result["key_unique"] = int(

            df[primary_key].nunique()

        )

    return result

inspection_results = []

for name, df in datasets.items():

    inspection_results.append(

        inspect_dataframe(

            name,

            df,

            primary_keys[name],

        )

    )

inspection_results

[{'dataset': 'customers',
  'rows': 150,
  'columns': 6,
  'column_names': ['customer_id',
   'name',
   'gender',
   'age',
   'city',
   'signup_date'],
  'dtypes': {'customer_id': 'int64',
   'name': 'str',
   'gender': 'str',
   'age': 'int64',
   'city': 'str',
   'signup_date': 'str'},
  'missing_total': 0,
  'duplicate_rows': 0,
  'primary_key': 'customer_id',
  'key_missing': 0,
  'key_duplicates': 0,
  'key_unique': 150},
 {'dataset': 'products',
  'rows': 100,
  'columns': 4,
  'column_names': ['product_id', 'product_name', 'category', 'price'],
  'dtypes': {'product_id': 'int64',
   'product_name': 'str',
   'category': 'str',
   'price': 'int64'},
  'missing_total': 0,
  'duplicate_rows': 0,
  'primary_key': 'product_id',
  'key_missing': 0,
  'key_duplicates': 0,
  'key_unique': 100},
 {'dataset': 'orders',
  'rows': 300,
  'columns': 5,
  'column_names': ['order_id',
   'customer_id',
   'order_date',
   'payment_method',
   'order_status'],
  'dtypes': {'order_id': 'int6

In [365]:
#63. LLM에 전달할 안전한 구조 요약

safe_schema_summary = {
    "datasets": inspection_results,

    "relationships": {
        "orders.customer_id": {

            "references": "customers.customer_id",

            "unknown_count": len(unknown_customer_ids),
        },

        "order_items.order_id": {

            "references": "orders.order_id",

            "unknown_count": len(unknown_order_ids),
        },
        "order_items.product_id": {

            "references": "products.product_id",

            "unknown_count": len(unknown_product_ids),
        },

    },

}

print(safe_schema_summary)

{'datasets': [{'dataset': 'customers', 'rows': 150, 'columns': 6, 'column_names': ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date'], 'dtypes': {'customer_id': 'int64', 'name': 'str', 'gender': 'str', 'age': 'int64', 'city': 'str', 'signup_date': 'str'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'customer_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 150}, {'dataset': 'products', 'rows': 100, 'columns': 4, 'column_names': ['product_id', 'product_name', 'category', 'price'], 'dtypes': {'product_id': 'int64', 'product_name': 'str', 'category': 'str', 'price': 'int64'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'product_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 100}, {'dataset': 'orders', 'rows': 300, 'columns': 5, 'column_names': ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status'], 'dtypes': {'order_id': 'int64', 'customer_id': 'int64', 'order_date': 'str', 'payment_method': 'str', 'order_status'

나는 온라인 쇼핑몰 데이터 분석을 준비하고 있습니다.

아래 정보는 원본 행이 아니라 구조 점검 요약입니다.

[{'datasets': [{'dataset': 'customers', 'rows': 150, 'columns': 6, 'column_names': ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date'], 'dtypes': {'customer_id': 'int64', 'name': 'str', 'gender': 'str', 'age': 'int64', 'city': 'str', 'signup_date': 'str'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'customer_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 150}, {'dataset': 'products', 'rows': 101, 'columns': 4, 'column_names': ['product_id', 'product_name', 'category', 'price'], 'dtypes': {'product_id': 'int64', 'product_name': 'str', 'category': 'str', 'price': 'int64'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'product_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 101}, {'dataset': 'orders', 'rows': 301, 'columns': 5, 'column_names': ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status'], 'dtypes': {'order_id': 'int64', 'customer_id': 'int64', 'order_date': 'str', 'payment_method': 'str', 'order_status': 'str'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'order_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 301}, {'dataset': 'order_items', 'rows': 766, 'columns': 5, 'column_names': ['order_item_id', 'order_id', 'product_id', 'quantity', 'unit_price'], 'dtypes': {'order_item_id': 'int64', 'order_id': 'int64', 'product_id': 'int64', 'quantity': 'int64', 'unit_price': 'int64'}, 'missing_total': 0, 'duplicate_rows': 0, 'primary_key': 'order_item_id', 'key_missing': 0, 'key_duplicates': 0, 'key_unique': 766}], 'relationships': {'orders.customer_id': {'references': 'customers.customer_id', 'unknown_count': 1}, 'order_items.order_id': {'references': 'orders.order_id', 'unknown_count': 1}, 'order_items.product_id': {'references': 'products.product_id', 'unknown_count': 1}}}]

다음을 검토해 주세요.

1. 각 데이터셋에서 한 행이 무엇을 의미할 가능성이 높은지 설명해 주세요.

2. 주요 ID와 외래 키를 구분해 주세요.

3. 즉시 삭제하면 안 되는 결측치·중복 사례를 알려 주세요.

4. 추가로 점검할 구조 문제를 우선순위대로 제안해 주세요.

5. 실제 데이터에 없는 컬럼이나 관계를 가정하지 마세요.

6. 각 제안에 대해 확인할 pandas 코드의 목적만 설명해 주세요.


# 65. LLM 답변 검증표

| 검증 항목 | 확인 내용 |
|---|---|
| 컬럼 존재 | 실제 컬럼명과 일치하는가? |
| 한 행의 의미 | 데이터셋 역할과 일치하는가? |
| ID 해석 | 식별자를 수치 변수로 해석하지 않았는가? |
| 중복 해석 | 반복 가능한 외래 키를 오류로 판단하지 않았는가? |
| 결측 해석 | 업무 의미 확인 없이 삭제를 권하지 않았는가? |
| 관계 | 실제 연결 키와 일치하는가? |
| 개인정보 | 추가 원본 행 제공을 요구하지 않는가? |

종합: 7개 항목 모두 통과. 다만 5번(관계) 항목은 참고로, unit_price vs products.price 차이를 "점검 필요"로만 언급했고 실제 값 비교는 하지 않았다는 점을 명확히 해둡니다 — 이는 가정이 아니라 확인이 필요한 미검증 항목으로 남겨둔 것입니다.

# 66데이터 첫인상 보고서

## 1. 데이터 파일
| 데이터셋 | 한 행의 의미 | 행 | 열 | 주요 ID |
|---|---|---|---|---|
| customers | 고객 1명 | 150 | 6 | customer_id |
| products | 상품(SKU) 1개 | 101 | 4 | product_id |
| orders | 주문 1건 (헤더) | 301 | 5 | order_id |
| order_items | 주문 내 상품 라인 1개 | 766 | 5 | order_item_id |

## 2. 컬럼과 타입
- customers: customer_id(int64), name(str), gender(str), age(int64), city(str), signup_date(str)
- products: product_id(int64), product_name(str), category(str), price(int64)
- orders: order_id(int64), customer_id(int64), order_date(str), payment_method(str), order_status(str)
- order_items: order_item_id(int64), order_id(int64), product_id(int64), quantity(int64), unit_price(int64)
- 날짜 컬럼(signup_date, order_date)이 str로 되어 있어 파싱 필요

## 3. 결측치
- 4개 데이터셋 모두 missing_total: 0 (컬럼 단위 결측 없음)

## 4. 전체 중복
- 4개 데이터셋 모두 duplicate_rows: 0

## 5. 주요 ID 결측·중복
- 4개 PK 모두 key_missing: 0, key_duplicates: 0
- key_unique: customers 150 / products 101 / orders 301 / order_items 766 (각 행 수와 일치, 정상)

## 6. 숫자형 범위
- 아직 미확인 (age, price, unit_price, quantity의 min/max/음수·0 여부는 구조 점검 요약에 없음)

## 7. 범주형 값
- 아직 미확인 (gender, city, payment_method, order_status, category의 실제 값·표기 일관성은 구조 점검 요약에 없음)

## 8. 날짜 범위와 변환 실패
- 아직 미확인 (signup_date, order_date 파싱 성공 여부, 최소·최대 날짜는 구조 점검 요약에 없음)

## 9. 파일 간 참조 무결성
- orders.customer_id → customers.customer_id : unknown_count 1건
- order_items.order_id → orders.order_id : unknown_count 1건
- order_items.product_id → products.product_id : unknown_count 1건

## 10. 발견한 문제
- 세 관계 모두 참조 무결성 위반이 1건씩 존재 (부모 테이블에 없는 참조값)
- 이 3건이 서로 연관된 것인지(예: 동일 주문 계열에서 발생) 여부는 아직 미확인

## 11. 즉시 처리하지 않은 이유
- unknown_count 1건이 실제로 어떤 값인지, 어떤 원인(탈퇴/오타/데이터 유실 등)인지 원본 데이터 없이는 확정할 수 없음
- 삭제할 경우 매출·주문 집계에 영향을 줄 수 있어 원인 확인이 선행되어야 함

## 12. 다음 전처리 단계에서 결정할 내용
- FK unknown 3건: 유지/제외/별도 플래그 처리 여부 결정
- 날짜 컬럼 파싱 방식 및 이상치 처리 기준 확정
- 숫자형(age, price, unit_price, quantity) 이상치 기준 확정
- 범주형 값 표기 정규화 규칙 확정